### PR TITLE
Feature/edit more id3 tags

### PR DESCRIPTION
**Pull request by igorer88**
This PR helps to solve #732

- Added some basic metadata missing fields to edit _track field_.
- Fix artist description label that was on title field.

### CODE REVIEW

**Comment by martpie** [comment]
Ok, I'll start with your comments on #732:

- I fixed the DB imports, rebasing on `master` should make things work now

> The cover is present on the Other tag, while Museeks only searches on the Front Cover tag.

If we can scan for the cover in some kind of priority (`Front Cover` first, then `Other`, then then any cover), that would sounds good to me. 

> About storing the cover on the tag, well that's the standard way, it means that it would work on every other music player.

Ideally, it would be great to have the option to do both,. The reason I am a bit reticent for that is I am afraid of messing someone's library files, and that's why placing a `cover.jpg` in their folder seems safer, even though it's not x-player compatible. I am starting why some players copy all files to another location before changing them, but I would like to avoid doing that. 

Assuming a 512x512 should be <100KB, it's probably ok to embed them into the file itself. As the cover is not set in the DB anyway, I would do that separately from this PR though:

- 1 PR for more fields udpate
- 1 PR for optionally edit ID3 tags
- 1 PR for setting the cover as ID3

File: src/lib/utils-library.ts [hunk 1]
```diff
@@ -120,3 +120,10 @@ const SORT_ORDERS: Record<SortBy, SortConfig> = {
 export function getSortOrder(sortBy: SortBy): SortConfig {
   return SORT_ORDERS[sortBy];
 }
+
+/**
+ *  Formats the track's duration from seconds to hh:mm:ss
+ */
+export function formatDuration(seconds: number) {
+  return new Date(seconds * 1000).toISOString().substring(11, 19);
+}
```
Comment (martpie) [inline src/lib/utils-library.ts:129]: You can probably re-use that instead, as it's already in the codebase and is unit-tested:

https://github.com/martpie/museeks/blob/01ad926c002db3c11861ed8065aa20ca33b072cb/src/hooks/useFormattedDuration.ts#L3-L24

https://github.com/martpie/museeks/blob/01ad926c002db3c11861ed8065aa20ca33b072cb/src/hooks/__tests__/useFormattedDuration.test.ts#L5-L14

File: src/generated/typings/index.ts [hunk 2]
```diff
@@ -30,5 +30,5 @@ export type SortOrder = "Asc" | "Dsc";
 /**
  * Track
  * represent a single track, id and path should be unique
- */
-export type Track = { _id: string, title: string, album: string, artists: Array<string>, genres: Array<string>, year: number | null, duration: number, track: NumberOf, disk: NumberOf, path: string, };
+*/
+export type Track = { _id: string, title: string, album: string, artists: Array<string>, genres: Array<string>, year: number, duration: number, track: NumberOf, disk: NumberOf, path: string, };
```
Comment (martpie) [inline src/generated/typings/index.ts:34]: Why remove the `null` from track? This typing is generated via `cargo test` and mimics the definition in Rust:

https://github.com/martpie/museeks/blob/master/src-tauri/src/libs/track.rs#L18-L30

so if you don't keep them in sync the CI will scream at you.

File: src/views/ViewTrackDetails.tsx [hunk 3]
```diff
@@ -26,12 +28,28 @@ export default function ViewTrackDetails() {
...
+    >
   >({
     title: track.title ?? '',
     artists: track.artists,
     album: track.album ?? '',
     genres: track.genres,
+    year: track.year ?? 0,
+    duration: track.duration,
+    track: track.track ?? { no: 0, of: 0 },
+    disk: track.disk ?? { no: 0, of: 0 },
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:51]: I would be careful with those defaults. The real default should probably be `{ no: null, of: null }`. Also, `track.track` and `track.disk` should technically always be defined.

File: src/views/ViewTrackDetails.tsx [hunk 4]
```diff
@@ -62,7 +80,6 @@ export default function ViewTrackDetails() {
         <Setting.Section>
           <Setting.Input
             label="Title"
-            description="You can add multiple artists with commas"
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:65]: Good catch

File: src/views/ViewTrackDetails.tsx [hunk 5]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
             }}
           />
         </Setting.Section>
+        <Setting.Section>
+          <Setting.Input
+            label="Year"
+            id="year"
+            name="year"
+            type="number"
+            min="1900"
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:142]: You never know ;) classical music and all that. Min: 0 would probably be safer, even if highly improbable.

File: src/views/ViewTrackDetails.tsx [hunk 6]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
...
+            step="1"
+            value={formData.year}
+            onChange={(e) => {
+              setFormData({ ...formData, year: Number(e.currentTarget.value) });
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:146]: I think we need to be more careful here as `e` or `NaN` for example are considered a valid number in JS.

File: src/views/ViewTrackDetails.tsx [hunk 7]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
...
+            }}
+          />
+        </Setting.Section>
+        <Setting.Section>
+          <Setting.Input
+            label="Duration"
+            id="duration"
+            name="duration"
+            type="text"
+            pattern="([01]?[0-9]{1}|2[0-3]{1}):[0-5]{1}[0-9]{1}"
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:156]: Probably no need for a pattern here as it's never going to change. Maybe don't even include this field?

File: src/views/ViewTrackDetails.tsx [hunk 8]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
...
+            readOnly={true}
+            value={formatDuration(formData.duration)}
+            onChange={(e) => {
+              setFormData({
+                ...formData,
+                duration: Number(e.currentTarget.value),
+              });
+            }}
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:164]: `onChange` probably not needed as we are never going to edit this data.

File: src/views/ViewTrackDetails.tsx [hunk 9]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
...
+          />
+        </Setting.Section>
+        <Setting.Section>
+          <Setting.Input
+            label="Duration"
+            id="duration"
+            name="duration"
+            type="text"
+            pattern="([01]?[0-9]{1}|2[0-3]{1}):[0-5]{1}[0-9]{1}"
+            readOnly={true}
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:157]: I would suggest `disabled`, that way it won't receive focus, won't be send to the form event and will get appropriate styling

File: src/views/ViewTrackDetails.tsx [hunk 10]
```diff
@@ -26,12 +28,28 @@ export default function ViewTrackDetails() {
...
+      | 'track'
+      | 'disk'
+      | 'path'
+    >
   >({
     title: track.title ?? '',
     artists: track.artists,
     album: track.album ?? '',
     genres: track.genres,
+    year: track.year ?? 0,
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:48]: `null` by default please, those fields can be empty.

File: src/views/ViewTrackDetails.tsx (hunk 6 above)
Comment (martpie) [reply src/views/ViewTrackDetails.tsx:146]: Also, we should allow for empty values: pseudo code: `if (value === "") formData.year = null`

File: src/views/ViewTrackDetails.tsx [hunk 11]
```diff
@@ -115,6 +133,130 @@ export default function ViewTrackDetails() {
...
+              name="track"
+              type="number"
+              min="0"
+              step="1"
+              value={Number(formData.track.no)}
+              onChange={(e) => {
+                setFormData({
+                  ...formData,
+                  track: {
+                    no: Number(e.currentTarget.value),
```
Comment (martpie) [inline src/views/ViewTrackDetails.tsx:181]: Same for all these, they should allow empty values (that set the value to null), not 0 by default.

**Review by martpie — CHANGES_REQUESTED** [review]
Overall looks great, thank you! Left comments.

For the CI failures, if you are using VSCode, I suggest you install the recommended extensions: they will give you linting and formatting warnings + format on save: https://github.com/martpie/museeks/blob/master/.vscode/extensions.json

**Comment by martpie** [comment]
Thank you for the update, could you squash your commits and make sure master is the base branch? merge commits make it really hard to review as a the commits [from master are duplicated](https://github.com/martpie/museeks/pull/791/files).

**Comment by igorer88** [comment]
Ok, I did a rebase and I think I broke it further. hahahha

imagen

I did a bun install But it' s still missing that macro error.

imagen

**Comment by igorer88** [comment]
I'm starting to think that I should do a new fork hahahaha

**Comment by martpie** [comment]
> I did a bun install But it' s still missing that macro error.

Try `cargo Install --path src-tauri` from the root of the repo? And make sure you got the latest version of rustup setup?

**Comment by igorer88** [comment]
> Thank you for the update, could you squash your commits and make sure master is the base branch? merge commits make it really hard to review as a the commits [from master are duplicated](https://github.com/martpie/museeks/pull/791/files).

I've never done commit squash before, so let me investigate how to do it.

**Comment by martpie** [comment]
https://www.git-tower.com/learn/git/faq/git-squash is a good resource :)

**Comment by martpie** [comment]
If you use Github Desktop, it's fairly simple too: 

Screenshot 2024-10-06 at 12 46 36

Then I guess, fetch all, rebase from mater, and hopefully this shoud work. Force pushing to this branch is ok, or you can play with another branch to be safe.

**Comment by igorer88** [comment]
Hi! @martpie sorry, I've been busy lately, had to travel and stuff. 
I'll do the squash and then I'll do another merge for the latest updates.

File: src/lib/utils-library.ts (hunk 1 above)
Comment (igorer88) [reply src/lib/utils-library.ts:129]: Done

**Comment by igorer88** [comment]
hi! @martpie, the master branch is broken again? or maybe I need to check the last merge...

imagen

File: src/views/ViewTrackDetails.tsx (hunk 8 above)
Comment (igorer88) [reply src/views/ViewTrackDetails.tsx:168]: done

File: src/views/ViewTrackDetails.tsx (hunk 7 above)
Comment (igorer88) [reply src/views/ViewTrackDetails.tsx:160]: Agreed, done

File: src/views/ViewTrackDetails.tsx (hunk 3 above)
Comment (igorer88) [reply src/views/ViewTrackDetails.tsx]: fixed to individual fields `track_no`, `track_of`, `disk_no`, `disk_of`...

File: src/generated/typings/index.ts (hunk 2 above)
Comment (igorer88) [reply src/generated/typings/index.ts]: changed back as it was before.

**Comment by martpie** [comment]
Master should be passing(CI is), make sure you rebase and follow the setup instructions (bun install, bun tauri dev etc). Eventually, cargo test too if you are changing some types

**Comment by martpie** [comment]
btw, are you using Limux Mac or Windows?

**Comment by igorer88** [comment]
Mac

**Comment by martpie** [comment]
Let me try to wrap this up, I'm finalising 0.20.0, and I'm doing a lot of chore/code revamps introducing a lot of conflicts

**Pull request closed without merging**